File: src/ai/.x/templates/openai-asst-streaming-with-functions-js/FunctionFactory.js
class FunctionFactory {
  constructor() {
    this.functions = {};
  }

  addFunction(schema, fun) {
    this.functions[schema.name] = { schema: schema, function: fun };
  }

  getFunctionSchemas() {
    return Object.values(this.functions).map(value => value.schema);
  }

  getTools() {
    return Object.values(this.functions).map(value => {
      return {
        type: "function",
        function: value.schema
      };
    });
  }

  tryCallFunction(function_name, function_arguments) {
    const function_info = this.functions[function_name];
    if (function_info === undefined) {
      return undefined;
    }

    var result = function_info.function(function_arguments);
    process.stdout.write(`\rassistant-function: ${function_name}(${function_arguments}) => ${result}\n`);
    process.stdout.write('\nAssistant: ');

    return result;
  }
}

exports.FunctionFactory = FunctionFactory;

File: src/ai/.x/templates/includes/openai.asst.py/function_factory.py
class FunctionFactory:
    def __init__(self):
        self.functions = {}

    def add_function(self, schema, func):
        self.functions[schema['name']] = {'schema': schema, 'function': func}

    def get_function_schemas(self):
        return [value['schema'] for value in self.functions.values()]

    def get_tools(self):
        return [
            {"type": "function", "function": value["schema"]}
            for value in self.functions.values()
        ]

    def try_call_function(self, function_name, function_arguments):
        function_info = self.functions.get(function_name)
        if function_info is None:
            return None

        result = function_info['function'](function_arguments)
        print(f"\rassistant-function: {function_name}({function_arguments}) => {result}")
        print("\nAssistant: ", end='')

        return result
